src/service/wrapped-service.ts

```diff
@@ -0,0 +1,333 @@
+
+export interface UpdateWrappedParameters {
+  ids?: number[],
+}
+
+export default class WrappedService extends WithManager {
+  public async getWrappedForUser(userId: number): Promise<WrappedResponse | null> {
+    const entityManager = this.manager;
+
+    const wrapped = await entityManager.findOne(Wrapped, { where: { userId } });
+    if (!wrapped) return null;
+    return WrappedService.asWrappedResponse(wrapped as any);
+  }
+
+  private static asWrappedResponse(data: any): WrappedResponse {
+    const rawDate = data.transactionMaxDate ?? null;
+    let maxDateStr = '';
+    if (rawDate) {
+      if (typeof rawDate === 'string') maxDateStr = rawDate;
+      else if (rawDate instanceof Date) maxDateStr = rawDate.toISOString().slice(0, 10);
+      else {
+        try {
+          maxDateStr = new Date(rawDate).toISOString().slice(0, 10);
+        } catch (e) {
+          maxDateStr = '';
+        }
+      }
+    }
+
+    const formatDateTime = (rawDateTime: any): string => {
+      if (!rawDateTime) return '';
+      if (typeof rawDateTime === 'string') return rawDateTime;
+      if (rawDateTime instanceof Date) return rawDateTime.toISOString();
+      try {
+        return new Date(rawDateTime).toISOString();
+      } catch (e) {
+        return '';
+      }
+    };
+
+    return {
+      userId: data.userId,
+      transactions: {
+        transactionCount: Number(data.transactionCount ?? 0),
+        transactionPercentile: Number(data.transactionPercentile ?? 0),
+        transactionMaxDate: maxDateStr,
+        transactionMaxAmount: Number(data.transactionMaxAmount ?? 0),
+        transactionHeatmap: this.parseHeatmap(data.transactionHeatmap),
+      },
+      spentPercentile: Number(data.spentPercentile ?? 0),
+      syncedFrom: formatDateTime(data.syncedFrom),
+      syncedTo: formatDateTime(data.syncedTo),
+    } as WrappedResponse;
+  }
+
+  private static parseHeatmap(raw: any): number[] {
+    if (raw === undefined || raw === null) return [];
+    if (Array.isArray(raw)) return raw as number[];
+    try {
+      const parsed = typeof raw === 'string' ? JSON.parse(raw) : raw;
+      if (Array.isArray(parsed)) return parsed as number[];
+    } catch (e) {
+      // fallthrough
+    }
+    return [];
+  }
+
+  private static computePercentiles(values: { id: number; value: number }[]) {
+    const total = values.length;
+    if (total === 0) return new Map<number, number>();
+
+    const sorted = [...values].sort((a, b) => a.value - b.value);
+
+    const lessCountMap = new Map<number, number>();
+    for (let i = 0; i < sorted.length; i++) {
+      const value = sorted[i].value;
+      if (!lessCountMap.has(value)) {
+        lessCountMap.set(value, i);
+      }
+    }
+
+    const res = new Map<number, number>();
+    for (const entry of values) {
+      const less = lessCountMap.get(entry.value) ?? 0;
+      const pct = 100 * (1 - (less / total));
+      res.set(entry.id, Number(pct.toFixed(2)));
+    }
+    return res;
+  }
```

Comment:
Some comments could help, specifically some JSDOC (here and at multiple places) since it can be hard to follow what the intention of a function is.